### PR TITLE
#174797520 Reset password

### PR DESCRIPTION
**Pull request by florisng**
#### What does this PR do?
For the user, create a forgot password & reset password endpoints

#### Description of Task to be completed?
- Create Forgot password route
- Check if an email is provided
- Check if the provided email exists in database
- Send a link to reset the password to the given email
- Create Reset password route
- Check the token on the reset Password endpoint
- Create a password validation
- Check if the new password format is correct
- Check if the the new password and the confirm match
- Use bcrypt to hash the new password
- Update database with new password

#### How should this be manually tested?

- Clone the repo git clone `https://github.com/atlp-rwanda/script-heroes-bn-backend.git`
- Switch your working directory by running `cd script-heroes-bn-backend`
- Checkout the branch `git checkout ft-reset-password-#174797520`
- Open the folder with your code editor
- Run `npm install`
- Start your server
- Open postman
- Insert the localhost url on port 3000
- Send a PUT request using the forgot Password and in body, add your email:
```
{
       "email":"flongtest@gmail.com"
}
```
- Copie and paste the returned token on endpoint `{{url}}/api/resetPassword/token`
- Send a POST request using the reset Password with the following in body:
```
{
       "password": "BRAVO123",
       "confirmPassword": "BRAVO123"
}

```
Screenshot (30)

#### What are the relevant pivotal tracker stories?
[#174797520](https://www.pivotaltracker.com/story/show/174797520)

### CODE REVIEW

File: src/controllers/forgotPassword.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    if (!emailExist) {
+      return res
+        .status(409)
+        .json({ message: res.__('Email does NOT exist in our database !!!') });
+    }
+
+    // Save seconds from 1st January 1970
+    const token = jwt.sign({ id: User.id }, process.env.RESET_PASSWORD_KEY, {
+      expiresIn: '1440m'
+    });
```
Comment (elvisiraguha) [inline src/controllers/forgotPassword.js]: Don't use jwt like this, the are helper functions that handle this. Check this [file](https://github.com/atlp-rwanda/script-heroes-bn-backend/blob/develop/src/utils/jwtFunctions.js)

File: src/routes/api/index.js [hunk 2]
```diff
@@ -14,4 +18,14 @@ router.use((err, req, res, next) => {
   return next(err);
 });
 
+// Forgot password
+// const userForgot = new forgotPasswordController();
+// router.post('/forgotPassword', userForgot.forgot);
```
Comment (elvisiraguha) [inline src/routes/api/index.js]: Remove commented code

File: src/routes/api/index.js [hunk 3]
```diff
@@ -1,4 +1,8 @@
 import express from 'express';
+// import forgotPasswordController from '../../controllers/forgotPassword';
+// import resetPasswordController from '../../controllers/resetPassword';
```
Comment (elvisiraguha) [inline src/routes/api/index.js]: Remove commented code

File: src/routes/api/index.js [hunk 4]
```diff
@@ -14,4 +18,14 @@ router.use((err, req, res, next) => {
...
 });
 
+// Forgot password
+// const userForgot = new forgotPasswordController();
+// router.post('/forgotPassword', userForgot.forgot);
+router.post('/forgotPassword', forgot);
+
+// Reset password
+// const userReset = new resetPasswordController();
+// router.get('/resetPassword/:theToken', userReset.reset);
```
Comment (elvisiraguha) [inline src/routes/api/index.js]: Remove commented code

**Comment by elvisiraguha** [comment]
Kindly rebase

**Review by valens10 — CHANGES_REQUESTED** [review]
you need to rebase, your branch has conflicts with the base branch.

**Comment by valens10** [comment]
Your Travis is still failing, check out the issue in the Travis CLI job log and fix it.

File: src/controllers/forgotPassword.js (hunk 1 above)
Comment (florisng) [reply src/controllers/forgotPassword.js]: Now encode & decode are used.

File: src/routes/api/index.js (hunk 2 above)
Comment (florisng) [reply src/routes/api/index.js]: There are all removed now.

**Review by Adadie — COMMENTED** [review]
Would you
 try creating some tests, because the coverage is dropping a lot

File: package.json [hunk 5]
```diff
@@ -11,14 +11,12 @@
     "start": "npm run build && node dist/index",
     "dev": "nodemon --exec babel-node src/index.js",
     "build": "babel src --out-dir dist",
-    "test": "npm run pretest && nyc --reporter=html --reporter=text mocha --timeout 150000 -r @babel/register test/* --exit",
+    "test": "npm run pretest && nyc --testTimeout=30000 --reporter=html --reporter=text mocha -r @babel/register test/* --timeout 10000 --exit",
```
Comment (Habinezajanvier) [inline package.json]: Please remove a timeout which is doing nothing here

File: test/controllers/forgotPwd.test.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+      .send({ password: '11111111', confirmPassword: '22222222' })
+      .end((err, res) => {
+        if (err) done(err);
+        res.should.have.status(400);
+        res.body.should.be.a('object');
+        res.body.should.have.property('error').eq('The passwords do NOT match !!!');
+        done();
+      });
+  });
+});
```
Comment (Habinezajanvier) [inline test/controllers/forgotPwd.test.js]: Please add an empty line

**Review by Habinezajanvier — CHANGES_REQUESTED** [review]

**Comment by elvisiraguha** [comment]
Did you remember to translate your responses? and add swagger documentation?

**Review by dushimeemma — APPROVED** [review]
Please increase code coverage

File: locales/kin.json [hunk 7]
```diff
@@ -64,5 +64,11 @@
...
   "Profile is already complete": "Umwirondoro wawe uruzuye",
   "Profile completed succesfuly": "umwirondoro wujujwe neza",
-  "Profile updated succesfuly": "guhindura umwirondoro byagenze neza"
+  "Profile updated succesfuly": "guhindura umwirondoro byagenze neza",
+  "Reset password - Barefoot Nomad": "Guhindura umubare w'ibanga - Barefoot Nomad",
+	"Failed in sending Reset password email !!!": "Guhindura umubare w'ibanga ntibyakunze !!!",
+	"A link to reset your password has been sent to your email address !!!": "Ilinki yo guhindura umubare w'ibanga yoherejwe kuri imeri yanyu !!!",
+	"Email does NOT exist in our database !!!": "Imeri ntabwo iri muri sisitemu !!!",
+	"The passwords do NOT match !!!": "Imibare y'ibanga ntabwo isa !!!",
+	"Password well reset": "Guhindura umubare w'ibanga byagenze neza"
```
Comment (b0nbon1) [inline locales/kin.json]: the spacing on line 67-68 is different from 69-73

File: src/controllers/accomodation.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        where: {
+          id
+        }
+      }
+    );
+    return res
+      .status(200)
+      .json({ message: res.__('Accomodation successfully updated') });
+  }
+};
```
Comment (b0nbon1) [inline src/controllers/accomodation.js]: why are you adding accomodation and it's not part of this PR?

**Review by b0nbon1 — CHANGES_REQUESTED** [review]

**Review by Habinezajanvier — APPROVED** [review]

File: src/controllers/accomodation.js (hunk 8 above)
Comment (florisng) [reply src/controllers/accomodation.js]: It's because i rebased.

Comment (florisng) [reply src/controllers/accomodation.js:99]: Should i remove this file?

File: locales/kin.json (hunk 7 above)
Comment (florisng) [reply locales/kin.json]: I fixed that.

File: src/middlewares/validations/reset-checkPassword.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    password: Joi.string().min(8).required().messages({
+      'string.base': 'Password must be string',
+      'string.empty': 'Please fill in your Password',
+      'string.min': 'Password must be at least {#limit} characters long',
+      'any.required': 'Password is required'
+    }),
+    confirmPassword: Joi.string().min(8).required().messages({
+      'string.empty': 'Please confirm your Password',
+      'any.required': 'Confirm password is required'
+    })
```
Comment (ericus123) [inline src/middlewares/validations/reset-checkPassword.js]: Validate **passwords** properly. Any 8 white space characters are passing.Add validation patterns

Comment (florisng) [reply src/middlewares/validations/reset-checkPassword.js]: I added REGEX to make it strong.

File: src/helpers/emailTemplate.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+              border-radius: 20px;
+              box-shadow: 0px 0px 10px gray;
+            }
+            .btn {
+              background-color: #f9a826;
+              text-align: center;
+              padding: 10px;
+              border-radius: 10px;
+              text-decoration: none;
+              color: #000;
```
Comment (kipropbrian) [inline src/helpers/emailTemplate.js]: Can you rename this file, to something better. Name it more in line with what the actual file is doing

Comment (florisng) [reply src/helpers/emailTemplate.js]: I named it `linkToResetPasswordTemplate`

Comment (florisng) [reply src/helpers/emailTemplate.js]: I named it `linkToResetPasswordTemplate.js`

**Review by kipropbrian — CHANGES_REQUESTED** [review]
Use sendgrid and not nodemailer

**Review by b0nbon1 — APPROVED** [review]
Looks good
